### DISCUSSION TITLE
Clarify the motivation: what role do IDEs play?

### DISCUSSION THREAD

**Discussion by keyboardDrummer** [build-server-protocol/build-server-protocol#491] [Q&A]
In the motivation it says that the build server can also be used by IDEs:
> The Build Server Protocol aims to define common functionality that both build tools (servers) and language servers/**editors** (clients) understand.

However, as an IDE developer, I want to do as little work as possible per language that I want to support. Given that, and that I'm already connecting to an LSP server, I want the LSP server to provide me with everything I need to know. This includes information such as what test and build targets there are. If there are BSP servers for a language that can answer such questions, I would want the language's LSP server to connect to those BSP servers, and relay any relevant information to the IDE. 

For clarity's sake, I will restate my comment in a different way. Given IDEs, language servers, and build tools, if each IDE already knows which language server to use per language, and each language server knows which build tool to use per build format, then why would we also teach the IDE's which build tool to use per build format? That seems like an excess of knowledge that we're putting into the IDE.

Could you confirm in the motivation whether IDEs are supposed to connect to BSP servers or not?

If the answer is yes, then what is the scenario? One option would be that LSP currently is lacking APIs that you're already offering with BSP.

**Comment by jastice** [comment]
There's several reasonable approaches here:

1. build server provides build structure information and commands to language server via bsp. Here the build is transparent to the editor side.
2. bsp is implemented alongside lsp by build tool.
3. bsp as primary protocol between build tool and IDE. This is the use case for IntelliJ, which is essentially its own language server, but it requires mapping the build tool model to the IntelliJ model.

bsp aims to minimize the buildtool-specific implementation work required by an IDE or language server by offering a common protocol and structure that covers most of the ground

**Comment by olafurpg** [comment]
As I see it. From the client side, BSP abstracts over buildtool-specifics while LSP abstracts over language-specifics. On the server side, BSP abstracts over the IDE-specifics (whether language server or editor) while LSP abstracts over the editor specifics.

As @jastice points out, there are different IDE architectures where BSP might be useful. The motivation section could benefit from concrete examples of how BSP fits into the larger picture.

> One option would be that LSP currently is lacking API's that you're already offering with BSP.

There are some methods in BSP that might make sense in LSP like `buildTarget/test` and `buildTarget/run`. A language server can already maybe hack those features together via `textDocument/codeLens` and `window/logMessage` but a first-class representation might be better.

> Given IDE's, language servers, and build tools, if each IDE already knows which language server to use per language, and each language server knows which build tool to use per build format, then why would we also teach the IDE's which build tool to use per build format?

This is a good point. The architecture I have imagined for [Metals](https://scalameta.org/metals/) (a small language server I maintain) was to communicate with build tools without language client (editor) knowing about it.

**Reply by cameron-martin** [reply]
> A language server can already maybe hack those features together via textDocument/codeLens and window/logMessage but a first-class representation might be better.

If the BSP were to contain the location of the definition of the target in a `BuildTarget`, the language server wouldn't have to be involved in adding these "Build" and "Test" codelenses at all.

**Comment by jvican** [comment]
> I want the LSP server to provide me with everything I need to know. This includes information such as what test and build targets there are. If there are BSP servers for a language that can answer such questions, I would want the language's LSP server to connect to those BSP servers, and relay any relevant information to the IDE.
  
I think editors are interested in connecting to BSP servers on their own because they can use the build tool information to enhance the user experience. For example, by highlighting the base directories of the build targets in a project or showing their project name directly or, more generally, to fetch more projects' metadata that allow the IDEs to provide actions to the user that are not understood/formalized by LSP.

**Comment by PavelSosin** [comment]
I think that most important scenario is vscode front-end with eclipse/che 7 back-end. Build server is a sidecar of che workspace.NEXT i.e. docker image, regardless how it is implemented. It has direct access to workspace. EDI front-end can be written in any front-end language and  communicates with back-end using BSP protocol. Both front-end and back-end implementation are facilitated by official BSP libraries written in corresponding languages.

**Comment by lefou** [comment]
IMHO, There is also the polyglot aspect, which is currently a bit underspecified. Projects which use more that one languages often use a main build tool to orchestrate the build process, even if other tools will be used for specific tasks. BSP therefore also provides multiple language extensions. If BSP would only be consumed over LSP, than multiple LSP servers would try to start their own BSP server. Instead, a single BSP server should be used by multiple LSP servers. But to my understanding, there is no such mechanism anywhere implemented or specified. I think there are some missing pieces in the overall architecture.

**Comment by jastice** [comment]
> Instead, a single BSP server should be used by multiple LSP servers. But to my understanding, there is no such mechanism anywhere implemented or specified. I think there are some missing pieces in the overall architecture.

This would be my approach as well, but a BSP server may need to be designed to allow multiple connections. Aside from that, where do you see specification gaps?

**Reply by lefou** [reply]
There are probably no real specification gaps. It's all about implementations from here on. I think what's needed is a RFC, a Request for Adoption or some "Architecture Proposal" how BSP, LSP and IDEs should be used together. It needs to be pushed into existing projects.

**Reply by cameron-martin** [reply]
It depends whether you want the language servers to reuse a single instance of the build server. This would be useful if the build server stores a lot of cached state, for example.

**Reply by lefou** [reply]
For Mill, it would be ok to start a new instance of the build server when needed. (This works right now, although we will add some more safe guarding / target locking once this is a typical use case.) We keep most state on disc. Only speed might be an issue due to hot vs. cold JVM specifics, but we can that work out once it becomes a bottleneck.  What's essential to me is, that the server itself is based on the same build tool and thus the same meta model and setup.

**Reply by lefou** [reply]
But if it is, for some reasons, for LSP servers or IDEs easier to always talk to the same server, I'm sure we can build some generic multi-client server, that can manage multiple instance of BSP servers, that don't support multiple clients.
